llarp/CMakeLists.txt

```diff
@@ -160,280 +243,140 @@ add_library(lokinet-layer-onion
-  rpc/json_conversions.cpp
-  rpc/lokid_rpc_client.cpp
-  rpc/rpc_request_parser.cpp
-  rpc/rpc_server.cpp
-  rpc/endpoint_rpc.cpp
-)
-
-# optional peer stats library
-add_library(lokinet-peerstats
-  STATIC
-  peerstats/peer_db.cpp
-  peerstats/types.cpp
-)
-
-# kitchen sink to be removed after refactor
-add_library(lokinet-service-deprecated-kitchensink
-  STATIC
-  endpoint_base.cpp
-  exit/context.cpp
-  exit/endpoint.cpp
-  exit/exit_messages.cpp
-  exit/policy.cpp
-  exit/session.cpp
-  handlers/exit.cpp
-  handlers/tun.cpp
-  service/name.cpp
-  service/address.cpp
-  service/async_key_exchange.cpp
-  service/auth.cpp
-  service/convotag.cpp
-  service/context.cpp
-  service/endpoint_state.cpp
-  service/endpoint_util.cpp
-  service/endpoint.cpp
-  service/identity.cpp
-  service/info.cpp
-  service/intro_set.cpp
-  service/intro.cpp
-  service/lns_tracker.cpp
-  service/name.cpp
-  service/outbound_context.cpp
-  service/protocol.cpp
-  service/sendcontext.cpp
-  service/session.cpp
-  service/tag.cpp
-)
-
-
-# interal tooling for pybind
-add_library(lokinet-tooling INTERFACE)
-if(WITH_HIVE)
-  add_library(lokinet-hive-tooling
-    STATIC
-    tooling/router_hive.cpp
-    tooling/hive_router.cpp
-    tooling/hive_context.cpp
-  )
-  target_link_libraries(lokinet-tooling INTERFACE lokinet-hive-tooling)
-endif()
+add_library(lokinet-amalgum INTERFACE)
 
-# interface library for setting commone includes, linkage and flags.
+# interface library for setting common includes, linkage and flags.
 add_library(lokinet-base INTERFACE)
 target_include_directories(lokinet-base
   INTERFACE ${PROJECT_SOURCE_DIR} ${PROJECT_SOURCE_DIR}/include
 )
 target_link_libraries(lokinet-base INTERFACE quic)
 
-if(WITH_PEERSTATS)
-  target_compile_definitions(lokinet-base INTERFACE -DLOKINET_PEERSTATS_BACKEND)
-  target_link_libraries(lokinet-base INTERFACE sqlite_orm)
+# Link libraries to their internals
+target_link_libraries(lokinet-core-internal PUBLIC lokinet-utils lokinet-core-utils)
+target_link_libraries(lokinet-core PRIVATE lokinet-core-internal)
+# target_link_libraries(lokinet-rpc PRIVATE lokinet-rpc-internal)
+target_link_libraries(lokinet-wire PRIVATE lokinet-wire-internal)
+# target_link_libraries(lokinet-time-place PRIVATE lokinet-time-place-internal)
+target_link_libraries(lokinet-dns PRIVATE lokinet-dns-internal)
+target_link_libraries(lokinet-utils PRIVATE lokinet-utils-internal)
+
+# Link lokinet-dns to alternate libraries
+target_link_libraries(lokinet-dns PUBLIC lokinet-dns-platform)
+if (TARGET lokinet-dns-systemd)
+  target_link_libraries(lokinet-dns PUBLIC lokinet-dns-systemd)
 endif()
 
-# interface libraries for internal linkage
-add_library(lokinet-layers INTERFACE)
-add_library(lokinet-amalgum INTERFACE)
-
-# helper function to link a library to lokinet-base, enable lto, add to lokinet-amalgum and then link to other libs
-function(lokinet_link_lib libname)
-  message(DEBUG "created target: ${libname}")
-  enable_lto(${libname})
```

Comment:
This commit is currently dropping the `enable_lto` that gets applied to each library, but that needs to be brought back (LTO significantly reduces the size of release binaries, and optimizes performance as well).

We could do it in a different way, though.  Perhaps something something:

```cmake
# Add an internal lokinet static library target, enables LTO (if enabled) on the target,
# and links it to the common lokinet-base interface.
# Invoke with the target/library name (e.g. "lokinet-foo") and list of source files, e.g.
#     lokinet_add_library(lokinet-foo foo/source1.cpp foo/source2.cpp)
function(lokinet_add_library name)
    add_library(${libname} STATIC ${ARGN})
    target_link_libraries(${libname} PUBLIC lokinet-base)
    enable_lto(${libname})
endfunction()
```
and then changing all the `add_library(lokinet-foo STATIC sources...)` to `lokinet_add_library(lokinet-foo sources...)` to DRY out these two common additions.

Reply:
Cool, I will loop in `target_link_libraries(${libname} INTERFACE lokinet-amalgum)` as well